### DISCUSSION TITLE
Question about best persistance practices when using mikro orm on hexagonal architectures

### DISCUSSION THREAD

**Discussion by danielgak** [mikro-orm/mikro-orm#1382] [Q&A] [answered]
We've just created a new project appliying DDD practices to archive a hexagonal architecture, and as we use typescript and NestJS, we've decided to give mikro-orm a try.

In our way to create the boilerplate we've realized we have a little problem. As we dont want ORM-entities in our domain layer,  we will have mainly three types of objects:

- Dtos 
- Domain Models
- Entities

The comunication between the layers is archive permutating the object to something that is actually usefull that layer therefore the number of dependencies on each layer is minimal, an example of this will be a request to create a user that arrives to the backend as an dto, is transformed to a domain object and then is stored in a users table.

`Controller (user create dto) -> Domain (user domain object) -> Infrastructure (user mikro-orm entity)`

In this case persist and flush would work, as is a new entity and it have no problem when is saved.
`this.userRepository.persistAndFlush(userEntity);`

But **we are having trouble updating the entity** as with an hexagonal pattern, an update would first act as a find and then as a save/update:
 
`Controller (user update dto) -> Domain (no domain object) -> Infrastructure (user mikro-orm entity found!) -> 
  Domain (user domain object) (we do the app logic here) -> Infrastructure (save user mikro-orm entity)`

As we find the user entity, and then we tranform it back to a domain object... we are having trouble to save it back as we're getting a `UniqueConstraintViolationException`. When we transform back the User domain model to the User Entity, the perstist ORM method acts as is a new entity and when it tries to store it, it crashes into the same User that is actually stored.

TLDR: 
How is recommended to use persist/save/flush when we want to do all our app logic in models that are not MikroORM entities? Is there a way to do _upsert_? or somethign similar as _INSERT ON CONFLICT UPDATE_?

Some advice here will be also usefull! Thanks :)

**Comment by B4nan** [comment, marked as answer]
UoW requires you to mutate managed entity to issue update query. If you don't want to work with MikroORM entities in any way, you will lose pretty much all the power that makes MikroORM different - without UoW you will have to resolve to query builder. In recent versions the support for `on conflict` was added too. 

https://mikro-orm.io/docs/query-builder
https://github.com/mikro-orm/mikro-orm/blob/master/tests/QueryBuilder.test.ts#L1128

Personally I am not a big fan of this approach - I would let the ORM work with your entities to gain the most out of it. If you are concerned with decorators making your entities too coupled, you could use `EntitySchema`.

https://mikro-orm.io/docs/entity-schema#using-custom-entity-classes

Quite similar/related question was asked 20 min ago, so check that too: https://github.com/mikro-orm/mikro-orm/discussions/1380
